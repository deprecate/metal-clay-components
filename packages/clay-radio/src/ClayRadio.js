import Component from 'metal-component';
import defineWebComponent from 'metal-web-component';
import Soy from 'metal-soy';
import {Config} from 'metal-state';

import templates from './ClayRadio.soy.js';

/**
 * Implementation of the Metal Clay Radio.
 */
class ClayRadio extends Component {}

/**
 * State definition.
 * @type {!Object}
 * @static
 */
ClayRadio.STATE = {
	/**
	 * Flag to indicate if checkbox is checked.
	 * @instance
	 * @memberof ClayRadio
	 * @type {?bool}
	 * @default false
	 */
	checked: Config.bool().value(false),

	/**
	 * Flag to indicate if checkbox is disabled.
	 * @instance
	 * @memberof ClayRadio
	 * @type {?bool}
	 * @default false
	 */
	disabled: Config.bool().value(false),

	/**
	 * CSS classes to be applied to the element.
	 * @instance
	 * @memberof ClayRadio
	 * @type {?string|undefined}
	 * @default undefined
	 */
	elementClasses: Config.string(),

	/**
	 * Flag to indicate if label should be hide.
	 * @instance
	 * @memberof ClayRadio
	 * @type {?bool}
	 * @default false
	 */
	hideLabel: Config.bool().value(false),

	/**
	 * Id to be applied to the element.
	 * @instance
	 * @memberof ClayRadio
	 * @type {?string|undefined}
	 * @default undefined
	 */
	id: Config.string(),

	/**
	 * Flag to indicate if input should be wrapped to be inline or not.
	 * @instance
	 * @memberof ClayRadio
	 * @type {?bool}
	 * @default false
	 */
	inline: Config.bool().value(false),

	/**
	 * Label of the input. Required for accesibility.
	 * @instance
	 * @memberof ClayRadio
	 * @type {!string|html}
	 * @default undefined
	 */
	label: Config.any().required(),

	/**
	 * Name to be applied to the element.
	 * @instance
	 * @memberof ClayRadio
	 * @type {?string|undefined}
	 * @default undefined
	 */
	name: Config.string(),

	/**
	 * Value to be applied to the element.
	 * @instance
	 * @memberof ClayRadio
	 * @type {?string|undefined}
	 * @default undefined
	 */
	value: Config.string(),
};

defineWebComponent('clay-radio', ClayRadio);

Soy.register(ClayRadio, templates);

export {ClayRadio};
export default ClayRadio;
