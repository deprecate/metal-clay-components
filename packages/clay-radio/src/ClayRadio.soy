{namespace ClayRadio}

/**
 * This renders an icon element based on the Clay Checkbox definition.
 */
{template .render}
	{@param label: html | string}
	{@param? checked: bool}
	{@param? disabled: bool}
	{@param? elementClasses: string}
	{@param? hideLabel: bool}
	{@param? id: string}
	{@param? inline: bool}
	{@param? name: string}
	{@param? value: string}


	{let $attributes kind="attributes"}
		class="custom-control custom-radio
			{if $inline}
				{sp}custom-control-inline
			{/if}

			{if $elementClasses}
				{sp}{$elementClasses}
			{/if}
		"

		{if $id}
			id="{$id}"
		{/if}
	{/let}

	<div {$attributes}>
		{call .input}
			{param checked: $checked /}
			{param disabled: $disabled /}
			{param hideLabel: $hideLabel /}
			{param label: $label /}
			{param name: $name /}
			{param value: $value /}
		{/call}
	</div>
{/template}

/**
 * This renders the checkbox input
 **/
{template .input}
	{@param label: html | string}
	{@param? checked: bool}
	{@param? disabled: bool}
	{@param? hideLabel: bool}
	{@param? name: string}
	{@param? value: string}

	<label>
		{let $inputAttributes kind="attributes"}
			{if $checked}
				checked="checked"
			{/if}

			{if $disabled}
				disabled="disabled"
			{/if}

			class="custom-control-input"

			{if $name}
				name="{$name}"
			{/if}

			{if $value}
				value="{$value}"
			{/if}

			ref="input"
			type="radio"
			role="radio"
		{/let}

		<input {$inputAttributes}></input>

		<span class="custom-control-indicator"></span>

		{let $spanLabelClasses kind="text"}
			custom-control-description
			{if $hideLabel}
				{sp}sr-only
			{/if}
		{/let}

		<span class="{$spanLabelClasses}">{$label}</span>
	</label>
{/template}
