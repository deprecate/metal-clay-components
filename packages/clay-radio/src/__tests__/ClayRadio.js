import ClayRadio from '../ClayRadio';

let clayRadio;

describe('ClayRadio', function() {
	afterEach(() => {
		if (clayRadio) {
			clayRadio.dispose();
		}
	});

	it('should render default markup', () => {
		clayRadio = new ClayRadio({
			label: 'My Radio',
		});

		expect(clayRadio).toMatchSnapshot();
	});

	it('should render a radio with classes', () => {
		clayRadio = new ClayRadio({
			elementClasses: 'my-custom-class',
			label: 'My Radio',
		});

		expect(clayRadio).toMatchSnapshot();
	});

	it('should render a radio with id', () => {
		clayRadio = new ClayRadio({
			id: 'myId',
			label: 'My Radio',
		});

		expect(clayRadio).toMatchSnapshot();
	});

	it('should render a checked radio', () => {
		clayRadio = new ClayRadio({
			checked: true,
			label: 'My Radio',
		});

		expect(clayRadio).toMatchSnapshot();
	});

	it('should render a disabled radio', () => {
		clayRadio = new ClayRadio({
			disabled: true,
			label: 'My Radio',
		});

		expect(clayRadio).toMatchSnapshot();
	});

	it('should render a checked and disabled radio', () => {
		clayRadio = new ClayRadio({
			checked: true,
			disabled: true,
			label: 'My Radio',
		});

		expect(clayRadio).toMatchSnapshot();
	});

	it('should render a radio with hidden label', () => {
		clayRadio = new ClayRadio({
			hideLabel: true,
			label: 'My Radio',
		});

		expect(clayRadio).toMatchSnapshot();
	});

	it('should render an inline radio', () => {
		clayRadio = new ClayRadio({
			inline: true,
			label: 'My Radio',
		});

		expect(clayRadio).toMatchSnapshot();
	});

	it('should render a radio with name', () => {
		clayRadio = new ClayRadio({
			name: 'myradio',
			label: 'My Radio',
		});

		expect(clayRadio).toMatchSnapshot();
	});

	it('should render a radio with value', () => {
		clayRadio = new ClayRadio({
			label: 'My Radio',
			value: 'myradio',
		});

		expect(clayRadio).toMatchSnapshot();
	});
});
