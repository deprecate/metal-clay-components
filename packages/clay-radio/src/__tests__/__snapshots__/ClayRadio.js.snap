// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClayRadio should generate a checked and disabled radio 1`] = `
<div class="custom-control custom-radio">
  <label>
    <input checked="" disabled="disabled" class="custom-control-input" ref="input" type="radio" role="radio">
    <span class="custom-control-indicator"></span>
    <span class="custom-control-description">My Radio</span>
  </label>
</div>
`;

exports[`ClayRadio should generate a checked radio 1`] = `
<div class="custom-control custom-radio">
  <label>
    <input checked="" class="custom-control-input" ref="input" type="radio" role="radio">
    <span class="custom-control-indicator"></span>
    <span class="custom-control-description">My Radio</span>
  </label>
</div>
`;

exports[`ClayRadio should generate a disabled radio 1`] = `
<div class="custom-control custom-radio">
  <label>
    <input disabled="disabled" class="custom-control-input" ref="input" type="radio" role="radio">
    <span class="custom-control-indicator"></span>
    <span class="custom-control-description">My Radio</span>
  </label>
</div>
`;

exports[`ClayRadio should generate a radio with custom classes 1`] = `
<div class="custom-control custom-radio my-custom-class">
  <label>
    <input class="custom-control-input" ref="input" type="radio" role="radio">
    <span class="custom-control-indicator"></span>
    <span class="custom-control-description">My Radio</span>
  </label>
</div>
`;

exports[`ClayRadio should generate a radio with custom id 1`] = `
<div class="custom-control custom-radio" id="myRadio">
  <label>
    <input class="custom-control-input" ref="input" type="radio" role="radio">
    <span class="custom-control-indicator"></span>
    <span class="custom-control-description">My Radio</span>
  </label>
</div>
`;

exports[`ClayRadio should generate a radio with hidden label 1`] = `
<div class="custom-control custom-radio">
  <label>
    <input class="custom-control-input" ref="input" type="radio" role="radio">
    <span class="custom-control-indicator"></span>
    <span class="custom-control-description sr-only">My Radio</span>
  </label>
</div>
`;

exports[`ClayRadio should generate a radio with name 1`] = `
<div class="custom-control custom-radio">
  <label>
    <input class="custom-control-input" name="myradio" ref="input" type="radio" role="radio">
    <span class="custom-control-indicator"></span>
    <span class="custom-control-description">My Radio</span>
  </label>
</div>
`;

exports[`ClayRadio should generate a radio with value 1`] = `
<div class="custom-control custom-radio">
  <label>
    <input class="custom-control-input" value="myradio" ref="input" type="radio" role="radio">
    <span class="custom-control-indicator"></span>
    <span class="custom-control-description">My Radio</span>
  </label>
</div>
`;

exports[`ClayRadio should generate an inline radio 1`] = `
<div class="custom-control custom-radio custom-control-inline">
  <label>
    <input class="custom-control-input" ref="input" type="radio" role="radio">
    <span class="custom-control-indicator"></span>
    <span class="custom-control-description">My Radio</span>
  </label>
</div>
`;

exports[`ClayRadio should generate an inline radio with classes 1`] = `
<div class="custom-control custom-radio custom-control-inline my-custom-class">
  <label>
    <input class="custom-control-input" ref="input" type="radio" role="radio">
    <span class="custom-control-indicator"></span>
    <span class="custom-control-description">My Radio</span>
  </label>
</div>
`;

exports[`ClayRadio should generate an inline radio with custom id 1`] = `
<div class="custom-control custom-radio custom-control-inline" id="myRadio">
  <label>
    <input class="custom-control-input" ref="input" type="radio" role="radio">
    <span class="custom-control-indicator"></span>
    <span class="custom-control-description">My Radio</span>
  </label>
</div>
`;

exports[`ClayRadio should generate default markup 1`] = `
<div class="custom-control custom-radio">
  <label>
    <input class="custom-control-input" ref="input" type="radio" role="radio">
    <span class="custom-control-indicator"></span>
    <span class="custom-control-description">My Radio</span>
  </label>
</div>
`;
